### PR TITLE
Jakarta Tags 2.0

### PR DESCRIPTION
**Pull request by arjantijms**
Signed-off-by: arjantijms <arjan.tijms@gmail.com>

## Specification PR template
When creating a specification project release review, create PRs with the content defined as follows.

Include the following in the PR:
- [X] A directory in the form wombat/x.y where x.y is the release major.minor version, and the directory contains the following.
- [X] Specification PDF in the form of jakarta-wombat-spec-x.y.pdf
- [X] Specification HTML in the form of jakarta-wombat-spec-x.y.html
- [X] A specification page named _index.md following the template at:
      https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md
- [ ] For a Progress Review, that sufficient progress has been made on a Compatible Implementation and TCK, to ensure that the spec is implementable and testable.
- [X] For a [Release Review](https://www.eclipse.org/projects/handbook/#release-review), a summary that a Compatible Implementation is complete, passes the TCK, and that the TCK includes sufficient coverage of the specification. The TCK users guide MUST include the instructions to run the compatible implementations used to validate the release.
Instructions MAY be by reference.
   - [x] Updated release record
   - [x] Generated IP Log
   - [x] Email to PMC
   - [x] Start release review by emailing EMO 
- [x] The URL of the OSSRH staging repository for the api, javadoc:
      https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/servlet/jsp/jstl/jakarta.servlet.jsp.jstl-api/2.0.0/
- [x] The URL of the staging directory on downloads.eclipse.org for the proposed EFTL TCK binary:
      https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-tags-tck-2.0.0.zip
- [X] The URL of the compatibility certification request issue:
      https://github.com/eclipse-ee4j/jstl-api/issues/126
- [X] Specification JavaDoc in the wombat/x.y/apidocs directory. 
If desired, an optional second PR can be created to contain just the JavaDoc in the `apidocs` directory.

Note: If any item does not apply, check it and mark N/A below it.

### CODE REVIEW

**Comment by kwsutter** [comment]
Assigned @keilw as the Spec Committee mentor.  Will add the checklist while I'm at it.  Thanks, @arjantijms, for initiating this draft PR!

**Comment by kwsutter** [comment]
# Spec Review Checklist

1. Spec PR
  - [ ] PR uses [template](https://github.com/jakartaee/specifications/blob/master/pull_request_template.md)
  - [ ] Directory of form {spec}/x.y
  - [ ] PDF of form jakarta-{spec}-spec-x.y.pdf ("-spec" preferred but not required)
  - [ ] HTML of form jakarta-{spec}-spec-x.y.html ("-spec" preferred but not required)
  - [ ] Index page {spec}/x.y/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_page_template.md)
  - [ ] Index page {spec}/_index.md following [template](https://github.com/jakartaee/specification-committee/blob/master/spec_index_template.md)
  - [ ] No other files (e.g., no jakarta_ee_logo_schooner_color_stacked_default.png)
  - [ ] Staging repository link of the form https://jakarta.oss.sonatype.org/content/repositories/staging/jakarta/{spec}/jakarta.{spec}-api/x.y.z/
  - [ ] EFTL TCK link of the form http://download.eclipse.org/.../+.zip
  - [ ] Compatibility certification link of the form https://github.com/eclipse-ee4j/{project}/#{issue}
  - [ ] (Optional) Second PR for just apidocs

2. _index.md
  - [ ] Link to project release plan of the form https://projects.eclipse.org/projects/ee4j.{spec}/releases/x.y/plan
  - [ ] Link to spec pdf
  - [ ] Link to spec html
  - [ ] Link to apidocs
  - [ ] Link to final TCK download zip file of the form https://download.eclipse.org/jakartaee/{spec}/x.y/*{spec}-tck-x.y.z.zip (The folder path is required, the file name pattern is preferred.)
  - [ ] Link to API jar file of the form https://search.maven.org/artifact/jakarta.{spec}/jakarta.{spec}-api/x.y.z/jar
  - [ ] Name of and link to at least one Compatible Implementation

3. javadocs
  - [ ] Footer contains Eclipse copyright and link to license
  - [ ] ESFL license is included, usually as doc-files/speclicense.html
  - [ ] no META-INF directory in PR
  - [ ] javadocs-jar artifact matches apidocs (optional for this release)

4. Spec PDF
  - [ ] Correct spec title
  - [ ] Version number of the form x.y, not x.y.z
  - [ ] Correct Eclipse copyright line
  - [ ] No DRAFT or SNAPSHOT
  - [ ] Correct Logo

5. Spec HTML
  - [ ] Same as PDF

6. TCK zip file
  - [ ] README file (optional for this release)
  - [ ] EFTL license file, preferably named LICENSE.md
  - [ ] User's Guide (or equivalent documentation)
  - [ ] How to test the Compatible Implementation(s) listed in _index.md above with the TCK (may be in UG)

7. TCK User's Guide (or equivalent documentation)
  - [ ] Software requirements listed
  - [ ] Installation and configuration described
  - [ ] How to run tests
  - [ ] Where to file challenges

8. Compatibility certification request
  - [ ] Request follows [template](https://github.com/jakartaee/specification-committee/blob/master/compatibility-certification-request.md)
  - [ ] SHA-256 fingerprint matches staged TCK zip file
  - [ ] Request issue has `certification` label.

9. TCK results summary
  - [ ] Page is hosted by Compatible Implementation project
  - [ ] Includes all information from certification request
  - [ ] Summary includes number of tests passed, failed, errors
  - [ ] SHA-256 fingerprint matches staged TCK zip file on cert request
 
10. Update Jakarta EE API jar
  - [ ] Update the Jakarta EE API jar by submitting a PR to the jakartaee-api project that updates the version number of your API jar file.

**Comment by pnicolucci** [comment]
@arjantijms , question so I know in the future. It looks as though in the jstl-api project upon building the `apidocs/jquery` directory contains scripts used by the javadocs. However in the specification project for these pull requests those same scripts reside in `apidocs/script-dir`. Is there a manual find and replace step that you do to get the apidocs to use the script-dir vs the jstl-api jquery directory? Or is there a build command that you use?

**Comment by arjantijms** [comment]
TCK re-executed: https://ci.eclipse.org/jstl/job/2_jstl_run_tck_against_staged_build/13/artifact/tags-tck/bin/summary.txt

**Comment by arjantijms** [comment]
> In general is there another command other than mvn clean install -Dstatus=Final that is used to generate the apidocs for these PRs?

Nope, it's just `mvn clean install` at either the top level, or specifically from with the /api folder to only generate the java docs. Everything from `target/apidocs` is copied directly.

**Comment by keilw** [comment]
For the Spec the copyright notice says  "Copyright (c) 2018, 2020 Eclipse Foundation."
In the javadocs footer "Copyright © 2019, 2020 Eclipse Foundation."
Is that on purpose, or should it be the same, like e.g. Mail?

**Comment by arjantijms** [comment]
> Is that on purpose, or should the years match, like e.g. Mail?

TBH not 100% sure, but let's make them match.

**Comment by keilw** [comment]
I asked about the years and similar to Connectors they seem acceptable, but what about the META-INF?
And what about "The URL of the compatibility certification request issue:"?

**Comment by arjantijms** [comment]
> but what about the META-INF?
> And what about "The URL of the compatibility certification request issue:"?

I'll take a look at this now, thx!

**Comment by arjantijms** [comment]
> The spec license is there, the name is doc-files/EFSL.html, but I guess that's fine?

Yes, that should be fine. The "speclicense" name is more common, but the file name itself doesn't seem to matter.

**Comment by arjantijms** [comment]
> but what about the META-INF?

Can you provide a link to where you saw that one? I may be blind, but I can't find it so easily.

**Comment by arjantijms** [comment]
TCK Results PR for GlassFish: https://github.com/eclipse-ee4j/glassfish/pull/23246

I wanted to host them at the Jakarta Tags website too, but it has not website activated yet. Mail sent: https://eclipse.org/lists/jstl-dev/msg00052.html

**Comment by arjantijms** [comment]
Created bug to activate website: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567942

**Comment by arjantijms** [comment]
@keilw 

> And what about "The URL of the compatibility certification request issue:"?

I've added it: https://github.com/eclipse-ee4j/jstl-api/issues/126

**Comment by arjantijms** [comment]
IP Log/CQ: https://dev.eclipse.org/ipzilla/show_bug.cgi?id=22734

**Comment by arjantijms** [comment]
Mail to PMC: https://eclipse.org/lists/ee4j-pmc/msg02881.html

**Comment by arjantijms** [comment]
Release review: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567944

**Comment by arjantijms** [comment]
Updated the javadoc to the current date, and the spec docs with the previously unmerged changes.

**Comment by arjantijms** [comment]
@keilw Can you take a look again?

File: tags/2.0/_index.md [hunk 1]
```diff
@@ -10,17 +10,18 @@ existing custom tags with Jakarta Standard Tag Library tags.
...
+* [Jakarta Standard Tag Library 2.0 TCK](https://download.eclipse.org/jakartaee/tags/jakarta-tags-tck-2.0.0.zip)([sig](https://download.eclipse.org/jakartaee/tags/jakarta-tags-tck-2.0.0.zip.sig),[sha](https://download.eclipse.org/jakartaee/tags/jakarta-tags-tck-2.0.0.zip.sha256),[pub](https://raw.githubusercontent.com/jakartaee/specification-committee/master/jakartaee-spec-committee.pub))
 * Maven coordinates
-  * [jakarta.servlet.jsp.jstl:jakarta.servlet.jsp.jstl-api:jar:2.0.0](https://search.maven.org/artifact/jakarta.servlet.jsp.jstl/jakarta.servlet.jsp.jstl-api/2.0.0-RC1/jar)
+  * [jakarta.servlet.jsp.jstl:jakarta.servlet.jsp.jstl-api:jar:2.0.0](https://search.maven.org/artifact/jakarta.servlet.jsp.jstl/jakarta.servlet.jsp.jstl-api/2.0.0/jar)
 
 
 # Compatible Implementations
 
-* [Eclipse Glassfish 6.0.0-M1](https://projects.eclipse.org/projects/ee4j.glassfish/downloads)
+* [Eclipse Glassfish 6.0.0-SNAPSHOT-2020-09-20](https://download.eclipse.org/ee4j/glassfish/weekly/glassfish-6.0.0-SNAPSHOT-2020-09-20.zip)
```
Comment (kwsutter) [inline tags/2.0/_index.md:24]: This GF link no longer seems to work. And, it's not consistent with the Certification Request that points at a Tags 2.0 implementation...

**Review by kwsutter — COMMENTED** [review]
@arjantijms, I'll take a look and provide some comments...

- Your Eclipse Release Record needs a sentence or two about what's new or unique with this release of Tags.  The EMO has been indicating on other release records that major releases (such as Tags 2.0) need something more than just pointing at the Jakarta EE 9 release plan.  Thanks.

**Comment by kwsutter** [comment]
> @keilw Can you take a look again?

I took a look and checked off boxes in the checklist.  The only items I found were documented in my review comments.  The TCK SHA and CI link need to be cleaned up.  Otherwise, I think this is pretty much ready to go.  @keilw, if you can confirm, then maybe we can start the ballot soon?  Thanks!

**Comment by keilw** [comment]
I need to switch to this PR from the Connector one, but that is already on the ballot. I guess the mentioned issues in the MD have not been addressed yet, have they?

File: tags/2.0/_index.md (hunk 1 above)
Comment (arjantijms) [reply tags/2.0/_index.md]: Thanks, it's updated now!

**Comment by arjantijms** [comment]
> Your Eclipse Release Record needs a sentence or two about what's new or unique with this release of Tags. 

Yes, I agree. I updated it now to reflect more accurately what's new: https://projects.eclipse.org/projects/ee4j.jstl/releases/2.0.0

**Comment by arjantijms** [comment]
@keilw The MD issues should be addressed

**Comment by arjantijms** [comment]
@keilw @kwsutter can you take one more look? Can this one go to ballot now?

**Comment by arjantijms** [comment]
If everything is fine now, I'll also promote the TCK and update the links to the promoted location.

**Comment by keilw** [comment]
Looks good to me. Maybe if @kwsutter could have a final look in the morning, then I'll start the ballot before the Spec Committee call.

**Comment by arjantijms** [comment]
Since all checkboxes have been checked-off, I'll promote the TCK now.

**Comment by arjantijms** [comment]
TCK promoted: https://download.eclipse.org/ee4j/jakartaee-tck/jakartaee9-eftl/promoted/jakarta-tags-tck-2.0.0.zip

**Comment by keilw** [comment]
Then I will also start the ballot.

**Review by kwsutter — APPROVED** [review]

**Comment by arjantijms** [comment]
@keilw I think we can close the ballot as the 14 day period has passed ;)

**Comment by keilw** [comment]
Ballot complete!

Few more steps now. Spec committee mentor (myself) need to:
- [ ] adds this final checklist to the main PR.
- [ ] adds the approved label to the PRs, and sends out the Ballot Summary per this template to the public Jakarta EE Specification Committee email list
- [ ] calculates the staged EFTL TCK signature and promotes it to the committee download area using the https://ci.eclipse.org/jakartaee-spec-committee/job/promote-release/ job.
- [ ] updates the specification page with the ballot results.
- [ ] merges the specification (and apidocs) PRs, ensuring the "date:" field in the _index.md file has an appropriate value to allow publishing.
- [ ] notifies the EMO of the ballot results by email to emo@eclipse-foundation.org
- [ ] creates an issue in the specification project that includes the following checklist for the specification project team:
  - [ ] promotes api staging release promotes the specification api jars to maven central. An example release job script can be found here https://wiki.eclipse.org/MavenReleaseScript.
  - [ ] go through the merged jakarta.ee specification website page to verify all the links are valid.
  - [ ] approve the compatibility request.
  - [ ] The compatible implementation project/vendor MUST send an email to tck@eclipse.org for approval of the compatible implementation for trademark usage.
  - [ ] merge any final release branch as appropriate for the branch management for the project.

**Comment by arjantijms** [comment]
Ballot concluded successfully post: https://www.eclipse.org/lists/jakarta.ee-spec/msg01276.html

**Comment by arjantijms** [comment]
Release scheduled for 11 November now: https://bugs.eclipse.org/bugs/show_bug.cgi?id=567944

**Pull request merged**